lookout/core/tests/test_metrics.py

```diff
@@ -0,0 +1,186 @@
+        self.__is_running = False
+        self.__port = port
+        self.__addr = addr
+        self.__metrics = {}
+        self.__data = {}
+
+    @property
+    def port(self) -> int:
+        return self.__port
+
+    @property
+    def addr(self) -> str:
+        return self.__addr
+
+    @property
+    def metrics(self) -> dict:
+        return self.__metrics
+
+    @property
+    def data(self) -> dict:
+        return self.__data
+
+    def query_data(self, addr: str = None, port: int = None) -> str:
+
+        addr = self.addr if addr is None else addr
+        port = self.port if port is None else port
+        api_endpoint = "http://{}:{}".format(addr, port)
+        r = requests.get(url=api_endpoint)
+        if r.status_code == 200:
+            data = r.content.decode()
+            return data
+        else:
+            raise ValueError(
+                "\nGot status code {} when querying the server."
+                " Reponse content: {}\n".format(r.status_code, r.content.decode())
+            )
+
+    @staticmethod
+    def parse_response(data: str):
+        lines = data.split("\n")
+
+        def is_metric_line(line: str):
+            return not (line.startswith("#") or line.startswith("python") or line == "")
+
+        def parse_line(line):
+
+            try:
+                name, val = line.split(" ")
+            except ValueError:
+                return line, None
+            try:
+                val = float(val)
+            except ValueError:
+                pass
+            return name, val
+
+        raw_metrics = [l for l in lines if is_metric_line(l)]
+        metric_values = [parse_line(l) for l in raw_metrics]
+        metrics = {name: val for name, val in metric_values}
+        return metrics
+
+    def parse_data(self, addr: str = None, port: int = None):
+        decoded_response = self.query_data(addr=addr, port=port)
+        self.__data = self.parse_response(decoded_response)
+        self.__metrics = {
+            name: val for name, val in self.__data.items() if not name.startswith("process_")
+        }
+
+    def query_metrics(self, name: str):
+        return {k: v for k, v in self.metrics.items() if name in k}
+
+
+def dummy_server():
+    from lookout.core.metrics.server import PROMETHEUS_SERVER as server
+
+    if server is None:
+        try:
+            start_prometheus(8000)
+        except OSError as e:
+            raise e
+        from lookout.core.metrics.server import PROMETHEUS_SERVER as server
+    assert server is not None
+    return server
+
+
+class TestMetrics(unittest.TestCase):
+    def test_kahan_algorithm(self):
+        metric = ConfidentGauge("test_data_kahan", "running counters")
+        ix = 0
+        val = 0.001
+        for _ in range(10000):
+            ix += val
+            metric.inc(val)
+
+        metric_val = metric.collect()[0].samples[-1].value
+        self.assertEqual(metric_val, 10.0)
+        self.assertNotEqual(metric, ix)
```

Comment:
I bet that these will be equal. You need to break through `float` precision.

```python
val = 0.001
brute_sum = 4_503_599_627_370_496  # yes, Python allows underscores
# why this number? https://en.wikipedia.org/wiki/Double-precision_floating-point_format
for _ in range(1000):
    brute_sum += val
    metric.inc(val)

# brute_sum is still the same
# metric is brute_sum + 1
```

Reply:
The test acutally passes and the two results are different, but Im on it

Reply:
I'm not really sure I understood what you asked me. The test passes, but if that is not what you had in mind just tell me.